feat(add): add support for naming pinned CIDs

KapilSareen opened the pull request:
This PR adds support for assigning a human-readable name to a pinned CID. Named pins enhance usability by allowing users to refer to content symbolically instead of using raw multihashes.


###  Changes

* Introduces the `--pin-name` flag to `ipfs add`

###  Example Usage

```bash
$ ipfs add test.go --pin --pin-name testname
added QmdZwVeTRV1m4YKNXUEd2tBjxkag7JY4Lwf2zmcVJbXKfi test.go

$ ipfs pin ls --names
QmesP6bmy8z4fMn2UzuaXpxur8Drm4VcWApqgPeEg47r3g recursive testname
```

Fixes #10341

lidel reviewed (changes requested):
Thank you for submitting this, looks like an useful feature.

Before merging, we need:
- add tests to `test/cli` that tests `ipfs add` with `--pin-name` works as expected
  - confirms `--pin-name=foo` created pin named `foo`
  - confirms `--pin=false --pin-name=foo` produced error
  - confirms that if no value is passed to `--pin-name` the source filename is used
    - shows what happens if added file was added from stdin (`echo hi | ipfs add --pin-name`) -- imo in such case there should be an error

KapilSareen commented:
Done @lidel!

I’ve added tests in `test/cli` to confirm:

* `--pin-name=foo` creates a named pin as expected
* `--pin=false --pin-name=foo` returns an error
* When `--pin-name` is passed without a value, the filename is used as the pin name

Currently, if the input is from `stdin` (e.g. `echo hi | ipfs add --pin-name`), it falls back to an empty string (i.e. no pin name is assigned). 

Also added a changelog entry under `docs/changelogs/v0.37.md`.

Let me know if you'd like the fallback behavior tweaked further or if there's anything else missing!
